src/components/button/Button.js

```diff
@@ -1,30 +1,160 @@
+
 class Button extends Component {
+    wrapperRef = React.createRef();
+    enterAnimation = false;
+
+    state = {
+        feedback: null,
+        feedbackOutcome: null,
+    };
+
+    componentDidMount() {
+        this.maybeHandleFeedbackChange();
+    }
+
+    componentDidUpdate(prevProps) {
+        this.maybeHandleFeedbackChange(prevProps.feedback);
+    }
+
+    componentWillUnmount() {
+        this.clearFeedbackOutcomeTimers();
+    }
+
     render() {
         const { variant, disabled, children, fullWidth, ...rest } = this.props;
+        const { feedback, feedbackOutcome } = this.state;
+        const loading = feedback === 'loading';
+        const hasFeedback = feedback === 'loading' || feedback === 'success' || feedback === 'error';
+        const finalDisabled = disabled || loading;
         const finalClassName = classNames(
             styles.button,
             styles[variant],
             fullWidth && styles.fullWidth,
+            styles[feedbackOutcome],
+            loading && styles.loading,
+            hasFeedback && styles.feedback,
+        );
+
+        const wrapperClassName = classNames(
+            styles.buttonWrapper,
+            styles[feedbackOutcome],
+            styles[variant],
+        );
+
+        const propsFeedback = this.props.feedback;
+        const hasPropsFeedback = propsFeedback === 'loading' || propsFeedback === 'success' || propsFeedback === 'error';
+        const successBlockClassName = classNames(
+            hasPropsFeedback && styles.feedback,
+            styles.successBlock
+        );
+        const errorBlockClassName = classNames(
+            hasPropsFeedback && styles.feedback,
+            styles.errorBlock
         );
 
         return (
-            <button { ...rest } disabled={ disabled } className={ finalClassName } >
-                <span>{ children }</span>
-            </button>
+            <div className={ wrapperClassName } ref={ this.wrapperRef }>
+                <button { ...rest } disabled={ finalDisabled } className={ finalClassName } >
+                    <span>{ children }</span>
+                </button>
+                <ProgressBar
+                    running={ loading }
+                    className={ styles.progressBar }
+                    onBegin={ this.handleProgressBarBegin }
+                    onEnd={ this.handleProgressBarEnd } />
+
+                <span className={ successBlockClassName } onTransitionEnd={ this.handleSpanTransitionEnd }>
+                    <CheckmarkIcon className={ styles.checkmark } onTransitionEnd={ this.handleSuccessIconTransitionEnd } />
+                </span>
+                <span className={ errorBlockClassName } onTransitionEnd={ this.handleSpanTransitionEnd }>
+                    <CrossIcon className={ styles.cross } onTransitionEnd={ this.handleErrorIconTransitionEnd } />
+                </span>
+            </div>
         );
     }
+
+    maybeHandleFeedbackChange(prevFeedback) {
```

Comment:
`maybe` seems kind of strange 😕should we think of a better name?

I think we should call it `handleFeedbackChange`. The verification would be done in `componentDidUpdate` only since the `componentDidMount` actually wants to "handle" the change, right?

Reply:
Personally I think so too. This is part of the logic that was borrowed from discussify and is named as such because there isn't necessarily a feedback change when this function is called. 

I've moved this verification to the functions that called `maybeHandleFeedbackChange` and renamed it as suggested